src/lib.rs

```diff
@@ -1,14 +1,363 @@
+        0
+    }
+
+    /// Constructs a `MerkleTree` and populates it with the provided elements.
+    /// * `elements` - array of `Hash` elements used to populate the tree.
+    pub fn build<H: Hash>(elements: &[H]) -> MerkleTree {
+        let capacity = elements.len().next_power_of_two();
+        let padding = capacity - elements.len();
+        let padding_vec = vec![MerkleTree::pad(); padding];
+
+        // Level 0 hashes
+        let leaves = elements
+            .iter()
+            .map(hash_single)
+            .chain(padding_vec)
+            .collect();
+
+        let mut levels = Vec::new();
+        generate_tree_levels(&leaves, &mut levels);
+
+        MerkleTree {
+            levels,
+            capacity,
+            padding,
+        }
+    }
+
+    /// Returns the height of the tree.
+    pub fn height(&self) -> usize {
+        self.levels.len()
+    }
+
+    /// Creates a `MerkleProof` for a given index.
+    /// * `index` - index value to generate the proof for.
+    pub fn get_proof(&self, index: usize) -> MerkleProof {
+        let is_invalid_index = index >= self.len();
+        if is_invalid_index || self.is_empty() {
+            return MerkleProof::Invalid;
+        }
+
+        let mut nodes: Vec<u64> = Vec::new();
+
+        for level_n in 0..self.levels.len() - 1 {
+            let ancestor = ancestor_index(index, level_n);
+            let proof_node_index = sibling_index(ancestor);
+            nodes.push(self.levels[level_n][proof_node_index]);
+        }
+
+        MerkleProof::Proof {
+            nodes,
+            index,
+            root: self.root().expect("Non-empty trees always have a root"),
+        }
+    }
+
+    /// Returns the root of the tree. If the tree is empty, the root will be `None`.
+    pub fn root(&self) -> Option<u64> {
+        if self.is_empty() {
+            return None;
+        }
+        self.levels.get(self.height() - 1)?.first().copied()
+    }
+
+    /// Returns the capacity of the tree.
+    /// The capacity is the amount of space it has allocated.
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    /// Returns the length of the tree.
+    /// The length of the tree is the amount of elements it contains.
+    /// It may be different from the capacity.
+    pub fn len(&self) -> usize {
+        match self.levels.first() {
+            Option::None => 0_usize,
+            Option::Some(level) => level.len() - self.padding,
+        }
+    }
+
+    /// Returns wether a tree has no elements or not.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns wether a tree is at full capacity or not.
+    pub fn is_full(&self) -> bool {
+        self.padding == 0
+    }
+
+    /// Duplicates the capacity of a given tree by generating a new
+    /// equally sized tree and merging it to the original. Re-computing
+    /// the needed hashes.
+    /// This makes the tree one point taller and creates a new root.
```

Comment:
I don't understand what `merging` means in this context. Does this function add a new root where its children are the root of the old tree and the root of the new one?

Reply:
I improved the explanation of the algorithm.